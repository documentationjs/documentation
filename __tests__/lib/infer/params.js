const parse = require('../../../src/parsers/javascript');
const inferParams = require('../../../src/infer/params');

function toComment(fn, file) {
  return parse(
    {
      file,
      source: fn instanceof Function ? '(' + fn.toString() + ')' : fn
    },
    {}
  )[0];
}

function evaluate(fn, file) {
  return inferParams(toComment(fn, file));
}

test('mergeTrees', function() {
  expect(
    inferParams.mergeTrees(
      [],
      [
        {
          title: 'param',
          description: 'First arg!',
          name: 'a',
          type: {
            type: 'NameExpression',
            name: 'string'
          }
        }
      ]
    )
  ).toMatchSnapshot();

  expect(
    inferParams.mergeTrees(
      [
        {
          title: 'param',
          name: '$0',
          anonymous: true,
          parameterIndex: 0,
          type: {
            type: 'NameExpression',
            name: 'object'
          },
          properties: [
            {
              title: 'param',
              name: '$0.a',
              parameterIndex: 0,
              type: {
                type: 'NameExpression',
                name: 'string'
              },
              properties: []
            }
          ]
        }
      ],
      [
        {
          title: 'param',
          description: 'First arg!',
          name: 'a',
          type: {
            type: 'NameExpression',
            name: 'object'
          }
        }
      ]
    )
  ).toMatchSnapshot();
});

test('inferParams', function() {
  expect(
    evaluate(function() {
      /** Test */
      function f(x) {}
    }).params
  ).toEqual([{ lineNumber: 3, name: 'x', title: 'param' }]);

  expect(
    evaluate(function() {
      /** Test */
      const f = function(x) {};
    }).params
  ).toEqual([{ lineNumber: 3, name: 'x', title: 'param' }]);

  expect(
    evaluate(`/** Test */function f({ x, ...xs }) {};`).params
  ).toMatchSnapshot();

  expect(
    evaluate(`/** Test */function f([a: string, b, {c}]) {};`).params
  ).toMatchSnapshot();

  expect(
    evaluate(
      `
    /**
     * Test
     * @param {Object} a renamed destructuring param
     */
    var f = function({ x }) {};
  `
    ).params
  ).toMatchSnapshot();

  expect(evaluate('/** Test */ var f = (x) => {}').params).toEqual([
    { lineNumber: 1, name: 'x', title: 'param' }
  ]);

  expect(
    evaluate(function() {
      const x = 1;
      const g = function(y) {};
      /** Test */
      const f = function(x) {};
    }).params
  ).toEqual([{ lineNumber: 8, name: 'x', title: 'param' }]);

  expect(
    evaluate(
      `
    /** Test */
    function f(x = 4) {}
  `
    ).params
  ).toMatchSnapshot();

  expect(
    evaluate(
      `
    /** Test
     * @param {number} x
    */
    function f(x = 4) {}
  `
    ).params
  ).toMatchSnapshot();

  expect(
    evaluate(
      `
    /** Test */
    function f({ x: y }) {}
  `
    ).params
  ).toMatchSnapshot();

  expect(
    evaluate(
      `
    /** Test */
    function f({ x: { y: { z } } }) {}
  `
    ).params
  ).toMatchSnapshot();

  expect(evaluate('/** Test */ export function f(x) {}').params).toEqual([
    { lineNumber: 1, name: 'x', title: 'param' }
  ]);

  expect(
    evaluate('/** Test */ export default function f(x) {}').params
  ).toEqual([{ lineNumber: 1, name: 'x', title: 'param' }]);

  expect(
    evaluate(function() {
      /**
       * @class
       * @hideconstructor
       */
      function SomeClass(foo, bar) {}
    }).params
  ).toEqual([]);

  expect(
    evaluate(`
      /**
       * Test
       */
      class SomeClass {
        /**
         * @hideconstructor
         */
        constructor(foo, bar) {}
      }
    `).params
  ).toEqual([]);
});
