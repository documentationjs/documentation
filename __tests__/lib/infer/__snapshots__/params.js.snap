// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inferParams 1`] = `
Array [
  Object {
    "anonymous": true,
    "name": "$0",
    "properties": Array [
      Object {
        "lineNumber": 1,
        "name": "$0.x",
        "title": "param",
      },
      Object {
        "lineNumber": 1,
        "name": "$0.xs",
        "title": "param",
        "type": Object {
          "type": "RestType",
        },
      },
    ],
    "title": "param",
    "type": Object {
      "name": "Object",
      "type": "NameExpression",
    },
  },
]
`;

exports[`inferParams 2`] = `
Array [
  Object {
    "anonymous": true,
    "name": "$0",
    "properties": Array [
      Object {
        "lineNumber": 1,
        "name": "$0.0",
        "title": "param",
        "type": Object {
          "name": "string",
          "type": "NameExpression",
        },
      },
      Object {
        "lineNumber": 1,
        "name": "$0.1",
        "title": "param",
      },
      Object {
        "anonymous": true,
        "name": "$0.2",
        "properties": Array [
          Object {
            "lineNumber": 1,
            "name": "$0.2.c",
            "title": "param",
          },
        ],
        "title": "param",
        "type": Object {
          "name": "Object",
          "type": "NameExpression",
        },
      },
    ],
    "title": "param",
    "type": Object {
      "name": "Array",
      "type": "NameExpression",
    },
  },
]
`;

exports[`inferParams 3`] = `
Array [
  Object {
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 28,
                  "line": 1,
                  "offset": 27,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "text",
              "value": "renamed destructuring param",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 28,
              "line": 1,
              "offset": 27,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "paragraph",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 28,
          "line": 1,
          "offset": 27,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "lineNumber": 2,
    "name": "a",
    "properties": Array [
      Object {
        "lineNumber": 6,
        "name": "a.x",
        "title": "param",
      },
    ],
    "title": "param",
    "type": Object {
      "name": "Object",
      "type": "NameExpression",
    },
  },
]
`;

exports[`inferParams 4`] = `
Array [
  Object {
    "default": "4",
    "lineNumber": 3,
    "name": "x",
    "title": "param",
    "type": undefined,
  },
]
`;

exports[`inferParams 5`] = `
Array [
  Object {
    "default": "4",
    "lineNumber": 1,
    "name": "x",
    "title": "param",
    "type": Object {
      "name": "number",
      "type": "NameExpression",
    },
  },
]
`;

exports[`inferParams 6`] = `
Array [
  Object {
    "anonymous": true,
    "name": "$0",
    "properties": Array [
      Object {
        "lineNumber": 3,
        "name": "$0.x",
        "title": "param",
      },
    ],
    "title": "param",
    "type": Object {
      "name": "Object",
      "type": "NameExpression",
    },
  },
]
`;

exports[`inferParams 7`] = `
Array [
  Object {
    "anonymous": true,
    "name": "$0",
    "properties": Array [
      Object {
        "lineNumber": 3,
        "name": "$0.x.y.z",
        "title": "param",
      },
    ],
    "title": "param",
    "type": Object {
      "name": "Object",
      "type": "NameExpression",
    },
  },
]
`;

exports[`mergeTrees 1`] = `
Object {
  "errors": Array [
    Object {
      "commentLineNumber": undefined,
      "message": "An explicit parameter named a was specified but didn't match inferred information ",
    },
  ],
  "mergedParams": Array [
    Object {
      "description": "First arg!",
      "name": "a",
      "title": "param",
      "type": Object {
        "name": "string",
        "type": "NameExpression",
      },
    },
  ],
}
`;

exports[`mergeTrees 2`] = `
Object {
  "errors": Array [],
  "mergedParams": Array [
    Object {
      "description": "First arg!",
      "name": "a",
      "properties": Array [
        Object {
          "name": "a.a",
          "parameterIndex": 0,
          "properties": Array [],
          "title": "param",
          "type": Object {
            "name": "string",
            "type": "NameExpression",
          },
        },
      ],
      "title": "param",
      "type": Object {
        "name": "object",
        "type": "NameExpression",
      },
    },
  ],
}
`;
